### PR TITLE
Ai refactoring

### PR DESCRIPTION
**Pull request by stohrendorf**

### CODE REVIEW

**Review by serishema — COMMENTED** [review]
@stohrendorf i've done my best to point out anything that seems off. 
I got a pen and paper and checked out the step and drop limits and I think you're right. 
This was pretty challenging for me, I actually failed the AI class I did so sorry if i've missed something. 

I'd like to test the game and see what effect these changes have on the AI, which I should hopefully be able to do tommorrow. 
Hopefully i've given you something to get you started.

**Comment by serishema** [comment]
I played the first 3 levels. 
These are the things I noticed about enemy behaviour. 

* bats still "bounce" off the ground after dying from being shot.
  This is consistent and occurs every time.
* I was able to get a wolf stuck on a slant block that is lara's waist height. 
screenshot in this screenshot you can't see the wolf past lara's head, but it's there.
* in caves before the door at the end the wolves get stuck on the side of the stairs and don't attack up the stairs like
  they normally do. This is one box high (ie lara can just hang from the edge without touching the ground). 
  I couldn't find a specific drop/step limit in wolf.h/wolf.cpp so I assume it's a quater sector inherrited from AIAgent.
screenshot again the wolf is behind lara
* In City of Valabamca the bear at the end can get up the stairs to the 2nd level which i'm pretty sure it's not supposed be able to.
screenshot
* In Lost Valley, raptors can get onto the temple platform and into the temple. Possibily because each step is not larger than the step limit? 
screenshot

File: src/engine/objects/aiagent.h [hunk 1]
```diff
@@ -86,7 +86,7 @@ class AIAgent : public ModelObject
     m_creatureInfo->rotateHead(angle);
   }
 
-  bool animateCreature(const core::Angle& deltaRotationY, const core::Angle& tilt);
+  bool animateCreature(const core::Angle& collisionRotationY, const core::Angle& tilt);
```
[the remark this answers is not in the document]
Comment (stohrendorf) [reply src/engine/objects/aiagent.h:89]: It's only the name to better reflect the meaning.

File: src/engine/objects/aiagent.cpp [hunk 2]
```diff
@@ -149,12 +126,13 @@ bool AIAgent::animateCreature(const core::Angle& deltaRotationY, const core::Ang
...
+  const auto bboxBottom = m_state.location.position.Y + bbox.y.max;
 
   auto currentSector = m_state.location.moved(0_len, bbox.y.max, 0_len).updateRoom();
 
-  if(currentSector->box == nullptr || boxFloor - currentSector->box->floor > pathFinder.step
-     || boxFloor - currentSector->box->floor < pathFinder.drop
+  // fix location in case the entity moved to an invalid location, including checks for step/drop limits
+  if(currentSector->box == nullptr || currentBoxFloor < currentSector->box->floor - pathFinder.step
+     || currentBoxFloor > currentSector->box->floor - pathFinder.drop
      || m_state.getCurrentBox().get()->*zoneRef != currentSector->box->*zoneRef)
```
[the remark this answers is not in the document]
Comment (stohrendorf) [reply src/engine/objects/aiagent.cpp:136]: Zones are basically groups of boxes. There's one "zone category" for flying entities, and two for walking ones, depending on whether their step limit is 256 or not.

File: src/engine/objects/aiagent.cpp [hunk 3]
```diff
@@ -185,158 +163,182 @@ bool AIAgent::animateCreature(const core::Angle& deltaRotationY, const core::Ang
...
-        moveX = collisionBounds.min - inSectorX;
+        moveX = inSectorBounds.min - inSectorX;
       }
-      else if(moveZ == 0_len && cannotMoveTo(origin - testDz - testDx))
+      else if(moveZ == 0_len && cannotMoveTo(bottom - testDz - testDx))
       {
         switch(axisFromAngle(m_state.rotation.Y))
         {
         case core::Axis::Deg180:
-          [[fallthrough]];
```
[the remark this answers is not in the document]
Comment (stohrendorf) [reply src/engine/objects/aiagent.cpp:236]: The `[[fallthrough]];` is basically only to silence compiler warnings if you have code in a `case` block that's missing a `break` in certain code paths. In this case, these annotations only blow up the code size, as one code block matches two conditions, and I think it's less noisy to remove these annotations and easier to understand that two `case`s match one code block.

File: src/engine/ai/pathfinder.h [hunk 4]
```diff
@@ -30,10 +30,15 @@ struct PathFinder
 
   [[nodiscard]] bool canVisit(const world::Box& box) const noexcept;
 
-  //! @brief Movement limits
+  //! @brief Movement limits.
+  //! @warning Step and drop are negated.
```
[the remark this answers is not in the document]
Comment (stohrendorf) [reply src/engine/ai/pathfinder.h:34]: I try to write self-documenting code, but these two variables are certainly the most confusing ones in the whole code base.

**Comment by stohrendorf** [comment]
* Bats bouncing off the floor when falling is OG behavior. The AI is handling dead bats just like living entities, which means the gravity is working against the bats penetrating the floor, which causes a correction. See the lines near the comments `fly movements ended up penetrating the floor` and `fly up from the floor`.
* Wolves have a drop limit of `-1024` and their default step limit of `256`, see `gameflows/tr1/object_infos.py`, so this is a bug as the AI should have calculated a box path from the lower end of the steps. Which also means the wolf on the slanted sector should be able to just jump down from it. This means that somewhere the drop/step limit checks aren't applied properly.
* Raptors being able to enter the temple is fine, your explanation is correct. I've also seen the bear in City of Vilcabamba get up there for quite some time now, so if that's a bug, it's caused by something outside of the changes of this PR. Both have their default stp/drop limits of +/-256.

File: src/engine/objects/aiagent.cpp [hunk 5]
```diff
@@ -149,12 +126,13 @@ bool AIAgent::animateCreature(const core::Angle& deltaRotationY, const core::Ang
...
-  const auto bboxMaxY = m_state.location.position.Y + bbox.y.max;
+  const auto bboxBottom = m_state.location.position.Y + bbox.y.max;
 
   auto currentSector = m_state.location.moved(0_len, bbox.y.max, 0_len).updateRoom();
 
-  if(currentSector->box == nullptr || boxFloor - currentSector->box->floor > pathFinder.step
-     || boxFloor - currentSector->box->floor < pathFinder.drop
+  // fix location in case the entity moved to an invalid location, including checks for step/drop limits
+  if(currentSector->box == nullptr || currentBoxFloor < currentSector->box->floor - pathFinder.step
+     || currentBoxFloor > currentSector->box->floor - pathFinder.drop
```
[the remark this answers is not in the document]
Comment (stohrendorf) [reply src/engine/objects/aiagent.cpp:136]: I added a comment line, if the comment is clarifying enough to your future self, feel free to either resolve this conversation or ask for more clarifications.

**Comment by stohrendorf** [comment]
One more terrifying thought... I seriously hope the box floor/ceiling heights aren't negated, too. But if I remember correctly, when I once exported the level geometry to OBJ, the boxes looked fine and matched the level geometry, but I'm not sure whether I negated the heights without questioning it enough.

**Comment by serishema** [comment]
Unfortunately I can't tell from the code. My guess would be that if the floor/ceiling heights were negated then these changes would have made all enemies behave very strangely which isn't happening. 

I'm happy with everything, except for the wolves getting stuck. Do you want to address that in a seperate PR, possibly in the future or deal with it here?

if it's to be addressed in the future i'm happy to approve these changes.

**Comment by stohrendorf** [comment]
I found the issue with the pathfinder, it's just another case of "reverse thinking".

**Review by serishema — APPROVED** [review]
Tested and the wolves now behave correctly.

**Pull request merged**